src/main/java/com/aws/greengrass/clientdevices/auth/connectivity/usecases/RecordConnectivityChangesUseCase.java

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.aws.greengrass.clientdevices.auth.connectivity.usecases;
+
+import com.aws.greengrass.clientdevices.auth.api.UseCases;
+import com.aws.greengrass.clientdevices.auth.connectivity.ConnectivityInformation;
+import com.aws.greengrass.clientdevices.auth.connectivity.HostAddress;
+import com.aws.greengrass.clientdevices.auth.connectivity.UpdateConnectivityInformationRequest;
+import com.aws.greengrass.clientdevices.auth.connectivity.UpdateConnectivityInformationResponse;
+
+import java.util.Set;
+import java.util.stream.Collectors;
+import javax.inject.Inject;
+
+/**
+ * Update Connectivity Information.
+ * </p>
+ * This use case constructs current connectivity information from
+ * connectivity providers and returns a diff (if any) since the
+ * last update.
+ */
+public class RecordConnectivityChangesUseCase implements
+        UseCases.UseCase<UpdateConnectivityInformationResponse, UpdateConnectivityInformationRequest, Exception> {
+    private final ConnectivityInformation connectivityInformation;
+
+    @Inject
+    public RecordConnectivityChangesUseCase(ConnectivityInformation connectivityInformation) {
+        this.connectivityInformation = connectivityInformation;
+    }
+
+    @Override
+    public UpdateConnectivityInformationResponse apply(UpdateConnectivityInformationRequest updateRequest) {
+        Set<HostAddress> previousConnectivityInfo = connectivityInformation.getAggregatedConnectivityInformation();
```

Comment:
Mmm seems like the update request only has a single source (meaning we would call this once per source change) so there are a few things we could do 

```
 Set<HostAddress> existingAddresses  = connectivityInformation.getHostAddresses(request.getSource());
connectivityInformation.recordAddressChanges(request.getSource(), request.getHostAddresses());

// Gets the address difference between the passed argument and the current state returns a `Diff`
Diff<HostAddresses> diff = connectivityInformation.getAddressDifference(existingAddresses);

return new UpdateConnectivityInformationResponse(diff.getAdded(), diff.getRemoved());
```

Again this is more for inspiration, I might be missing something here but that way the application layer is only stitching the domain and the domain objects contain all the logic of diffing

Reply:
Added TODO for this